File: resources/recipes/devops/feature-flags/feature_flags.js
// Lightweight feature flag service with boolean, percentage, user, and group rollouts.

import { createHash } from "crypto";

class FeatureFlags {
  constructor(config) {
    this.config = config;
  }

  isEnabled(flag, userId = null) {
    const rule = this.config[flag] ?? false;

    if (typeof rule === "boolean") return rule;
    if (typeof rule !== "object") return false;

    if (rule.percentage != null && userId) {
      return this.#hashBucket(userId, flag) < rule.percentage;
    }
    if (rule.users && userId) {
      return rule.users.includes(userId);
    }
    if (rule.groups) {
      return this.#checkGroups(rule.groups);
    }
    return false;
  }

  #hashBucket(userId, flag) {
    const hash = createHash("md5").update(`${flag}:${userId}`).digest("hex");
    return parseInt(hash.slice(0, 8), 16) % 100;
  }

  #checkGroups(groups) {
    return false;
  }
}

if (import.meta.url === `file://${process.argv[1]}`) {
  const flags = new FeatureFlags({
    newDashboard: true,
    betaSearch: { percentage: 10 },
    vipFeature: { users: ["user_123"] },
    adminTools: { groups: ["admins"] },
  });

  console.log("newDashboard:", flags.isEnabled("newDashboard"));
  console.log("betaSearch (user_456):", flags.isEnabled("betaSearch", "user_456"));
  console.log("vipFeature (user_123):", flags.isEnabled("vipFeature", "user_123"));
  console.log("missing_flag:", flags.isEnabled("missing_flag"));
}

export { FeatureFlags };

File: resources/recipes/devops/feature-flags/feature_flags.py
"""Lightweight feature flag service with boolean, percentage, user, and group rollouts."""

from __future__ import annotations

import hashlib
from typing import Any


class FeatureFlags:
    def __init__(self, config: dict[str, Any]):
        self.config = config

    def is_enabled(self, flag: str, user_id: str | None = None) -> bool:
        rule = self.config.get(flag, False)

        if isinstance(rule, bool):
            return rule

        if isinstance(rule, dict):
            if "percentage" in rule and user_id:
                return self._hash_bucket(user_id, flag) < rule["percentage"]
            if "users" in rule and user_id:
                return user_id in rule["users"]
            if "groups" in rule:
                return self._check_groups(rule["groups"])

        return False

    def _hash_bucket(self, user_id: str, flag: str) -> int:
        digest = hashlib.md5(f"{flag}:{user_id}".encode()).hexdigest()
        return int(digest, 16) % 100

    def _check_groups(self, groups: list[str]) -> bool:
        # Hook for group membership lookup
        return False


if __name__ == "__main__":
    flags = FeatureFlags({
        "new_dashboard": True,
        "beta_search": {"percentage": 10},
        "vip_feature": {"users": ["user_123"]},
        "admin_tools": {"groups": ["admins"]},
    })

    print(f"new_dashboard: {flags.is_enabled('new_dashboard')}")
    print(f"beta_search (user_456): {flags.is_enabled('beta_search', user_id='user_456')}")
    print(f"vip_feature (user_123): {flags.is_enabled('vip_feature', user_id='user_123')}")
    print(f"missing_flag: {flags.is_enabled('missing_flag')}")
